File: algorithms/Java/backtracking/Sudoku_Solver.java
                                                          // ***** SUDOKU SOLVER ALGORITHM******** \\ (CONTRIBUTED BY : SHAWAIZ SIDDIQUI (github: @Shawaiz22))


//THIS IS THE ALGORITHM USING BACKTRACKING FOR SOLVING SUDOKU. 

// SUDOKU IS A VERY POPULAR PUZZLE GAME IN WHICH WE HAVE A 9X9 GRID IN WHICH WE HAVE TO FILL DIGITS (1-9) IN SUCH A MANNER THAT NO DIGIT IS REPEATED IN THE SAME ROW, 
// OR IN THE SAME COLUMN OR THE INSIDE THE SMALL GRID ( THE SMALL GRID IS A 3x3 GRID, WHICH IS A PART OF THIS 9X9 GRID, THERE ARE SUCH 9 GRIDS IN THE BIGGER GRID). 
// SO OUR TASK IS TO FILL THE NUMBERS IN THE BLANK SPACES (SOME SPACES WILL ALREADY BE FILLED WHICH CAN'T BE CHANGED), SUCH THAT IT MATCHES ALL THE CONDITIONS. WE USE 
// BACKTRACKING ALGORITHM HERE, WHICH SOLVES THIS PROBLEM IN VARIOUS HIT AND TRIALS AND FINALLY GIVES THE OUTPUT AS THE RESULTANT SOLUTION.


//*****************ALGORITHM INFORMATION******************\\

// Time complexity: O (9^ (n*n)). For every unassigned index, there are 9 possible options so the time complexity is O (9^ (n*n))

//*********************************************************\\



public class Sudoku_Solver {

    public static boolean sudokuSolution(int[][] sudoku, int row, int col){


        // base case:
        if(row==9){     //case when we have checked all the rows and crosed the last row
            return true;
        }

        int nextRow=row, nextCol=col+1;

        if(col+1==9){
            nextRow=row+1;
            nextCol=0;
        }

        if(sudoku[row][col]!=0){  /// condition when my sudoku is not blank
            return sudokuSolution(sudoku,nextRow,nextCol);
        }

        for(int digit=1;digit<=9;digit++){


         if(isSafe(sudoku,row,col,digit)){
             sudoku[row][col]=digit;
             // if it is safe to place the digit, then place it and call the recursive function for next row and next col!

             if(sudokuSolution(sudoku,nextRow,nextCol)){
                 return true;     //if solution exists for the next row and col too, then return true to the previous levels in call stack
             }
             // if it is not true, then I will backtrack and set zero at current position, because due to the previous placement of digit
                // solution does not exists,
           
           // backtrack
             sudoku[row][col]=0;
         }

        }

//        else if no digit fits here, then it has no solution, so return false from here
        return false;
    }


    public static boolean isSafe(int[][] sudoku, int row, int col, int digit){

        // row wise check
        for(int i=0;i<9;i++){
            if(sudoku[row][i]==digit){
                return false;
            }
        }

        // col wise check
        for(int i=0;i<9;i++){
            if(sudoku[i][col]==digit){
                return false;
            }
        }

        // grid check
      // here we find the start row and start cols of the grid using the formula, then travesing in the grid(3x3), we check for the digits
      
        int sr = (row/3)*3;
        int sc = (col/3)*3;

        for(int i=sr;i<sr+3;i++){
            for(int j=sc;j<sc+3;j++){
                if(sudoku[i][j]==digit){
                    return false;
                }
            }
        }

        return true;
    }

    public static void printSudoku(int[][] sudoku){
        System.out.println("_____________________");
        for(int i=0;i<sudoku.length;i++){
            System.out.print("| ");
            for(int j=0;j<sudoku[0].length;j++){
                System.out.print(sudoku[i][j]+" ");
            }
            System.out.print("| ");
            System.out.println();
        }
        System.out.println("_____________________");
    }

    public static void main(String[] args) {
        int[][] sudoku = {{5,6,8,0,9,3,0,0,7},
                          {3,4,0,0,0,7,0,0,5},
                          {0,9,7,5,0,4,6,0,3},
                          {7,0,0,0,1,2,5,0,0},
                          {0,1,9,0,0,8,7,6,0},
                          {0,8,0,7,0,0,0,3,0},
                          {0,2,0,0,3,5,8,0,0},
                          {0,5,1,2,0,6,3,0,0},
                          {0,0,0,9,0,0,0,0,0}};

        boolean res = sudokuSolution(sudoku,0,0);

        if(res){
            System.out.println("SOLVED SUDOKU IS: ");
            printSudoku(sudoku);
        }
        else{
            System.out.println("SOLUTION DOES NOT EXIST");
        }
    }
}

 


//************************************ (CODE CONTRIBUTED BY :  SHAWAIZ SIDDIQUI (github: @Shawaiz22) )******************************************\\

File: algorithms/CPlusPlus/Backtracking/sudoku-Solver.cpp
/*
Code Contributed by MOHIT KUMAR SUMAN to MakeContributions/DSA Github repository

Below the cpp code displays all the possible Solutions of Sudoku Solver

Problem : Given a paritally filled 9X9 2D array 'grid[9][9]', the goal is to assign digits (from 1 to 9) to the empty cells so that every row, column, and subgrid of size 3X3 contains exactly one instance of the digits from 1 to 9

Input:
grid = { {3, 0, 6, 5, 0, 8, 4, 0, 0}, 
         {5, 2, 0, 0, 0, 0, 0, 0, 0}, 
         {0, 8, 7, 0, 0, 0, 0, 3, 1}, 
         {0, 0, 3, 0, 1, 0, 0, 8, 0}, 
         {9, 0, 0, 8, 6, 3, 0, 0, 5}, 
         {0, 5, 0, 0, 9, 0, 6, 0, 0}, 
         {1, 3, 0, 0, 0, 0, 2, 5, 0}, 
         {0, 0, 0, 0, 0, 0, 0, 7, 4}, 
         {0, 0, 5, 2, 0, 6, 3, 0, 0} }
Output:
          3 1 6 5 7 8 4 9 2
          5 2 9 1 3 4 7 6 8
          4 8 7 6 2 9 5 3 1
          2 6 3 4 1 5 9 8 7
          9 7 4 8 6 3 1 2 5
          8 5 1 7 9 2 6 4 3
          1 3 8 9 4 7 2 5 6
          6 9 2 3 5 1 8 7 4
          7 4 5 2 8 6 3 1 9
Explanation: Each row, column and 3*3 box of 
the output matrix contains unique numbers.

Time Complexity :- O(9^(n*n))
Space Complexity :- O(n*n)

Approach:
1) Sudoku can be solved by one by one assigning numbers to empty cells, Before assigning a number, check whether it is safe to assign or not.

2) Check that the number is not present in the current roe, current column and current 3X3 subgrid.

3) After checking for safety, assign the number, and recursively check whether the assignerd number leads to the solution or not.

5) if no then try next number for the current empty cell
*/
#include<iostream>
#include<bits/stdc++.h>
using namespace std;

//UNASSIGNED is used for empty cells
#define UNASSIGNED 0
// N is used for the size of sudoku grid
#define N 9

bool findUnassignedLocation(int grid[N][N], int &row, int &col){
   for(row = 0; row < N; row++){
      for(col = 0; col < N; col++){
         if(grid[row][col] == UNASSIGNED){
            return true;
         }
      }
   }
   return false;
}

// find that the element exist in 3X3 row or not
bool usedInBox(int grid[N][N], int boxStartRow, int boxStartCol, int num){
   for(int row = 0; row < 3; row++){
      for(int col = 0; col < 3; col++){
         if(grid[row + boxStartRow][col + boxStartCol] == num){
            return true;
         }
      }
   }
   return false;
}

// find that the element exist in the entire row or not
bool usedInRow(int grid[N][N], int row, int num){
   for(int col = 0; col < N; col++){
      if(grid[row][col] == num){
         return true;
      }
   }
   return false;
}


// find that the element exist in the entire col or not
bool usedInCol(int grid[N][N], int col, int num){
   for(int row = 0; row < N; row++){
      if(grid[row][col] == num){
         return true;
      }
   }
   return false;
}

bool isSafe(int grid[N][N], int row, int col, int num){
   return !usedInRow(grid, row, num) && !usedInCol(grid, col, num) && !usedInBox(grid, row-row%3, col-col%3, num) && grid[row][col] == UNASSIGNED;
}

bool SolveSudoku(int grid[N][N]){
   int row, col;
   // if there is no unassigned location then we are done
   if(!findUnassignedLocation(grid, row, col)){
      return true;
   }

   for(int i = 1; i <= 9; i++){
      if(isSafe(grid, row, col, i)){
         grid[row][col] = i;
         if(SolveSudoku(grid)){
            return true;
         }
         // backtrack
         grid[row][col] = UNASSIGNED;
      }
   }
   return false;
}

void printGrid(int grid[N][N]){
   for(int row = 0; row < N; row++){
      for(int col = 0; col < N; col++){
         cout<<grid[row][col]<<" ";
      }
      cout<<endl;
   }
}

int main(){
   int grid[N][N] = {{3,0, 6, 5, 0, 8, 4, 0, 0 },
                  {5, 2, 0, 0, 0, 0, 0, 0, 0 },
                  {0, 8, 7, 0, 0, 0, 0, 3, 1 },
                  {0, 0, 3, 0, 1, 0, 0, 8, 0 },
                  {9, 0, 0, 8, 6, 3, 0, 0, 5 },
                  {0, 5, 0, 0, 9, 0, 6, 0, 0 },
                  {1, 3, 0, 0, 0, 0, 2, 5, 0 },
                  {0, 0, 0, 0, 0, 0, 0, 7, 4 },
                  {0, 0, 5, 2, 0, 6, 3, 0, 0 } };
   if(SolveSudoku(grid) == true){
      printGrid(grid);
   }else{
      cout<<"No Solution Exists";
   }
   return 0;
}

// to compile
// g++ .\sudokuSolver.cpp
// .\a.exe



